types/reference/image_reference.go

```diff
@@ -0,0 +1,36 @@
+package reference
+
+import (
+	"fmt"
+
+	distreference "github.com/docker/distribution/reference"
+)
+
+// Parse parses the given references and return the repository and
+// tag (if present) from it. If there is an error during parsing, it will
+// return an error.
+func Parse(ref string) (string, string, error) {
+	distributionRef, err := distreference.ParseNamed(ref)
+	if err != nil {
+		return "", "", err
+	}
+
+	repository, tag := GetRepositoryAndTagFromNamedRef(distributionRef)
+	return repository, tag, nil
+}
+
+// GetRepositoryAndTagFromNamedRef returns a repository and tag given the specified reference.
+// This function is necessary as long as the docker "server" api make the distinction between repository
+// and tags.
+func GetRepositoryAndTagFromNamedRef(ref distreference.Named) (string, string) {
+	var tag string
+	if tagged, isTagged := ref.(distreference.NamedTagged); isTagged {
+		tag = tagged.Tag()
+	}
+	// This is in order to support a reference like :
+	// name@sha256:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
+	if digested, isDigested := ref.(distreference.Digested); isDigested {
+		tag = fmt.Sprintf("%s@%s", tag, digested.Digest().String())
```

Comment:
I would prefer not to introduce another string format which is a subset of the reference format. If we want to support both tag and digest in the Docker API we should send it as a full reference and not try and to sneak it in with the tag.